modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -235,6 +258,79 @@ private CompletableFuture<Object> processScanRetrieveBatchAction(ReadWriteScanRe
         });
     }
 
+    /**
+     * Process transaction finish request:
+     * <ol>
+     *     <li>Evaluate commit timestamp.</li>
+     *     <li>Run specific raft {@code FinishTxCommand} command, that will apply txn state to corresponding txStateStorage.</li>
+     *     <li>Send cleanup requests to all enlisted primary replicas.</li>
+     * </ol>
+     * This operation is NOT idempotent, because of commit timestamp evaluation.
+     *
+     * @param request Transaction finish request.
+     * @return future result of the operation.
+     */
+    private CompletableFuture<Object> processTxFinishAction(TxFinishReplicaRequest request) {
+        HybridTimestamp commitTimestamp = hybridClock.now();
+
+        List<String> aggregatedGroupIds = request.groups().values().stream().flatMap(List::stream).collect(Collectors.toList());
+
+        UUID txId = request.txId();
+
+        boolean commit = request.commit();
+
+        CompletableFuture<Object> chaneStateFuture = raftClient.run(
+                new FinishTxCommand(
+                        txId,
+                        commit,
+                        commitTimestamp,
+                        aggregatedGroupIds
+                )
+        );
+
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-17578
+        chaneStateFuture.thenRun(
+                () -> request.groups().forEach(
+                        (recipientNode, replicationGroupIds) -> txManager.cleanup(
```

Comment:
I believe there is no necessary to invoke other manager, because you can do all internally in replication manager.

Reply:
Nope, there's replicaService inside txManager and there's no such in PartitionReplicaListener. ReplicaService is used in order to send cleanupRequest to all enlisted replicas.